File: python/sample_search/rrt.py
'''
@file: rrt.py
@breif: RRT motion planning
@author: Winter
@update: 2023.1.17
'''
import os, sys
import math
import numpy as np

sys.path.append(os.path.abspath(os.path.join(__file__, "../../")))

from .sample_search import SampleSearcher
from utils import Env, Node

class RRT(SampleSearcher):
    '''
    Class for RRT motion planning.
    [1] Rapidly-Exploring Random Trees: A New Tool for Path Planning

    Parameters
    ----------
    start: tuple
        start point coordinate
    goal: tuple
        goal point coordinate
    env: Env
        environment
    max_dist: float
        Maximum expansion distance one step
    sample_num: int
        Maximum number of sample points
    goal_sample_rate: float
        heuristic sample

    Examples
    ----------
    >>> from utils import Map
    >>> from sample_search import RRT
    >>> start = (5, 5)
    >>> goal = (45, 25)
    >>> env = Map(51, 31)
    >>> planner = RRT(start, goal, env)
    >>> planner.run()
    '''
    def __init__(self, start: tuple, goal: tuple, env: Env, max_dist: float, 
        sample_num: int, goal_sample_rate: float=0.05) -> None:
        super().__init__(start, goal, env)
        # Maximum expansion distance one step
        self.max_dist = max_dist
        # Maximum number of sample points
        self.sample_num = sample_num
        # heuristic sample
        self.goal_sample_rate = goal_sample_rate
        # Sampled list
        self.sample_list = [self.start]

    def __str__(self) -> str:
        return "Rapidly-exploring Random Tree(RRT)"

    def plan(self):
        '''
        RRT motion plan function.

        Return
        ----------
        cost: float
            path cost
        path: list
            planning path
        '''
        # main loop
        for _ in range(self.sample_num):
            # generate a random node in the map
            node_rand = self.generateRandomNode()

            # visited
            if node_rand in self.sample_list:
                continue
            
            # generate new node
            node_new = self.getNearest(self.sample_list, node_rand)
            if node_new:
                self.sample_list.append(node_new)
                dist = self.dist(node_new, self.goal)
                # goal found
                if dist <= self.max_dist and not self.isCollision(node_new, self.goal):
                    self.goal.parent = node_new.current
                    self.goal.g = node_new.g + self.dist(self.goal, node_new)
                    self.sample_list.append(self.goal)
                    return self.extractPath(self.sample_list)
        return 0, None

    def run(self) -> None:
        '''
        Running both plannig and animation.
        '''
        cost, path = self.plan()
        self.plot.animation(path, str(self), cost, self.sample_list)

    def generateRandomNode(self) -> Node:
        '''
        Generate a random node to extend exploring tree.

        Return
        ----------
        node: Node
            a random node based on sampling
        '''
        if np.random.random() > self.goal_sample_rate:
            current = (np.random.uniform(self.delta, self.env.x_range - self.delta),
                    np.random.uniform(self.delta, self.env.y_range - self.delta))
            return Node(current, None, 0, 0)
        return self.goal

    def getNearest(self, node_list: list, node: Node) -> Node:
        '''
        Get the node from `node_list` that is nearest to `node`.

        Parameters
        ----------
        node_list: list
            exploring list
        node: Node
            currently generated node

        Return
        ----------
        node: Node
            nearest node 
        '''
        # find nearest neighbor
        dist = [self.dist(node, nd) for nd in node_list]
        node_near = node_list[int(np.argmin(dist))]

        # regular and generate new node
        dist, theta = self.dist(node_near, node), self.angle(node_near, node)
        dist = min(self.max_dist, dist)
        node_new = Node((node_near.current[0] + dist * math.cos(theta),
                        (node_near.current[1] + dist * math.sin(theta))),
                         node_near.current, node_near.g + dist, 0)
        
        # obstacle check
        if self.isCollision(node_new, node_near):
            return None
        return node_new

    def extractPath(self, closed_set):
        '''
        Extract the path based on the CLOSED set.

        Parameters
        ----------
        closed_set: list
            CLOSED set

        Return
        ----------
        cost: float
            the cost of planning path
        path: list
            the planning path
        '''
        node = closed_set[closed_set.index(self.goal)]
        path = [node.current]
        cost = node.g
        while node != self.start:
            node_parent = closed_set[closed_set.index(Node(node.parent, None, None, None))]
            node = node_parent
            path.append(node.current)

        return cost, path

File: matlab/sample_search/rrt.m
function [path, flag, cost, expand] = rrt(map, start, goal)
%%
% @file: rrt.m
% @breif: RRT motion planning
% @paper: Rapidly-Exploring Random Trees: A New Tool for Path Planning
% @author: Winter
% @update: 2023.1.30

%%
    % Maximum expansion distance one step
    max_dist = 0.5;
    % Maximum number of sample points
    sample_num = 10000;
    % heuristic sample
    goal_sample_rate = 0.05;
    % sample list
    sample_list  = [start, 0, start];
    % map size
    [x_range, y_range] = size(map);
    % resolution
    resolution = 0.1;
    
    path = [];
    flag = false;
    cost = 0;
    expand = [];
    
    % main loop
    for i=1: sample_num
        % generate a random node in the map
        node_rand = generate_node(goal_sample_rate, goal, x_range, y_range);

        % visited
        if loc_list(node_rand, sample_list, [1, 2])
            continue
        end

        % generate new node
        [node_new, success] = get_nearest(sample_list, node_rand, map, max_dist, resolution);
        if success
            sample_list = [node_new; sample_list];
            distance = dist(node_new(1:2), goal');
            
            % goal found
            if distance <= max_dist && ~is_collision(node_new(1:2), goal, map, max_dist, resolution)
                goal_ = [goal, node_new(3) + distance, node_new(1:2)];
                sample_list = [goal_; sample_list];
                flag = true;
                cost = goal_(3);
                break
            end
        end
    end
    
    
    if flag
        path = extract_path(sample_list, start);
        expand = sample_list;
    end
end

%%
function index = loc_list(node, list, range)
% @breif: locate the node in given list
    num = size(list);
    index = 0;
    if ~num(1)
        return
    else  
        for i=1:num(1)
            if isequal(node(range), list(i, range))
                index = i;
                return;
            end
        end
    end
end

function node = generate_node(goal_sample_rate, goal, x_range, y_range)
%breif: Generate a random node to extend exploring tree.
    if rand() > goal_sample_rate
        x = 0.5 + (x_range - 1) * rand();
        y = 0.5 + (y_range - 1) * rand();
        node = [x, y];
        return
    end
    
    node = goal;
    return
end

function [new_node, flag] = get_nearest(node_list, node, map, max_dist, resolution)
%@breif: Get the node from `node_list` that is nearest to `node`.
    flag = false;
    % find nearest neighbor
    dist_vector = dist(node_list(:, 1:2), node');
    [~, index] = min(dist_vector);
    node_near = node_list(index, :);

    % regular and generate new node
    distance = min(dist(node_near(1:2), node'), max_dist);
    theta = angle(node_near, node);
    new_node = [node_near(1) + distance * cos(theta), ...
                           node_near(2) + distance * sin(theta), ...
                           node_near(3) + distance, ...
                           node_near(1:2)];

    % obstacle check
    if is_collision(new_node(1:2), node_near(1:2), map, max_dist, resolution)
        return
    end
    flag = true;
end

function flag = is_collision(node1, node2, map, max_dist, resolution)
%@breif: Judge collision when moving from node1 to node2.
    flag = true;
    theta = angle(node1, node2);
    distance = dist(node1, node2');

      % distance longer than the threshold
      if (distance > max_dist)
          return
      end
        
      % sample the line between two nodes and check obstacle
      n_step = round(distance / resolution);
      for i=1:n_step
          x = node1(1) + i * resolution * cos(theta);
          y = node1(2) + i * resolution * sin(theta);
          if map(round(x), round(y)) == 2
              return
          end
      end
      
      flag = false;
end

function path = extract_path(close, start)
% @breif: Extract the path based on the CLOSED set.
    path  = [];               
    closeNum = length(close(:, 1));
    index = 1;
    while 1
        path = [path; close(index, 1:2)];
        if isequal(close(index, 1:2), start)   
            break;
        end
        for i=1:closeNum
            if isequal(close(i, 1:2), close(index, 4:5))
                index = i;
                break;
            end
        end
    end
end

 function theta = angle(node1, node2)
    theta = atan2(node2(2) - node1(2), node2(1) - node1(1));
 end